packages/core/src/RenderPipeline/PrimitiveChunkManager.ts

```diff
@@ -0,0 +1,57 @@
+import { Engine } from "../Engine";
+import { PrimitiveChunk } from "./PrimitiveChunk";
+import { SubPrimitiveChunk } from "./SubPrimitiveChunk";
+
+/**
+ * @internal
+ */
+export class PrimitiveChunkManager {
+  /** The maximum number of vertex. */
+  static MAX_VERTEX_COUNT = 4096;
+
+  primitiveChunks = new Array<PrimitiveChunk>();
+
+  constructor(
+    public engine: Engine,
+    public maxVertexCount = PrimitiveChunkManager.MAX_VERTEX_COUNT
+  ) {}
+
+  allocateSubChunk(vertexCount: number): SubPrimitiveChunk {
+    const primitiveChunks = this.primitiveChunks;
+    const length = primitiveChunks.length;
+    let subChunk: SubPrimitiveChunk = null;
+    for (let i = 0; i < length; ++i) {
+      subChunk = primitiveChunks[i].allocateSubChunk(vertexCount);
+      if (subChunk) {
+        subChunk.id = i;
+        return subChunk;
+      }
+    }
+
+    const data = (this.primitiveChunks[length] ||= new PrimitiveChunk(this.engine, this.maxVertexCount));
```

Comment:
const data = (primitiveChunks[length] ||= new PrimitiveChunk(this.engine, this.maxVertexCount));

Reply:
done